core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,839 @@
+          equal = self(i) == that(i)
+          i += 1
+        }
+
+        equal
+      }
+    case _ => false
+  }
+
+  /**
+   * Returns a filtered subset of this chunk.
+   */
+  def filter(f: A => Boolean): Chunk[A] = {
+    implicit val B = Chunk.classTagOf(this)
+
+    val len  = self.length
+    val dest = Array.ofDim[A](len)
+
+    var i = 0
+    var j = 0
+    while (i < len) {
+      val elem = self(i)
+
+      if (f(elem)) {
+        dest(j) = elem
+        j += 1
+      }
+
+      i += 1
+    }
+
+    if (dest == null) Chunk.Empty
+    else Chunk.Slice(Chunk.Arr(dest), 0, j)
+  }
+
+  /**
+   * Returns the concatenation of mapping every element into a new chunk using
+   * the specified function.
+   */
+  def flatMap[@specialized B](f: A => Chunk[B]): Chunk[B] = {
+    val len                    = self.length
+    var chunks: List[Chunk[B]] = Nil
+
+    var i               = 0
+    var total           = 0
+    var B0: ClassTag[B] = null.asInstanceOf[ClassTag[B]]
+    while (i < len) {
+      val chunk = f(self(i))
+      if (chunk.length > 0) B0 = Chunk.classTagOf(chunk)
+      chunks ::= chunk
+      i += 1
+      total += chunk.length
+    }
+    chunks = chunks.reverse
+
+    if (B0 == null) Chunk.empty
+    else {
+      implicit val B = B0
+
+      val dest: Array[B] = Array.ofDim(total)
+
+      val it = chunks.iterator
+      var n  = 0
+      while (it.hasNext) {
+        val chunk = it.next
+
+        chunk.toArray(n, dest)
+        n += chunk.length
+      }
+
+      Chunk.Arr(dest)
+    }
+  }
+
+  final def foldMLazy[E, S](z: IO[E, S])(pred: S => Boolean)(f: (S, A) => IO[E, S]): IO[E, S] = {
+    val len = length
+
+    def loop(s: S, i: Int): IO[E, S] =
+      if (i >= len) IO.now(s)
+      else {
+        if (pred(s)) f(s, self(i)).flatMap(loop(_, i + 1))
+        else IO.now(s)
+      }
+
+    z.flatMap(loop(_, 0))
+  }
+
+  final def foldLeftLazy[S](z: S)(pred: S => Boolean)(f: (S, A) => S): S = {
+    val len = length
+    var s   = z
+    var i   = 0
+    while (i < len && pred(s)) {
+      s = f(s, self(i))
+      i += 1
+    }
+    s
+  }
+
+  /**
+   * Folds over the elements in this chunk from the left.
+   */
+  def foldLeft[@specialized S](s0: S)(f: (S, A) => S): S = {
+    val len = self.length
+    var s   = s0
+
+    var i = 0
+    while (i < len) {
+      s = f(s, self(i))
+      i += 1
+    }
+
+    s
+  }
+
+  /**
+   * Effectfully folds over the elements in this chunk from the left.
+   */
+  final def foldM[E, S](s: S)(f: (S, A) => IO[E, S]): IO[E, S] =
+    foldLeft[IO[E, S]](IO.now(s)) { (s, a) =>
+      s.flatMap(f(_, a))
+    }
+
+  /**
+   * Folds over the elements in this chunk from the right.
+   */
+  def foldRight[@specialized S](s0: S)(f: (A, S) => S): S = {
+    val len = self.length
+    var s   = s0
+
+    var i = len - 1
+    while (i >= 0) {
+      s = f(self(i), s)
+      i -= 1
+    }
+
+    s
+  }
+
+  override final def hashCode: Int = toArray.toSeq.hashCode
```

Comment:
wouldn't `Array.hashCode(toArray.toSeq)` be faster?

Reply:
```scala
@ Array.hashCode(Array(1,2,3).toSeq)
cmd0.sc:1: no arguments allowed for nullary method hashCode: ()Int
val res0 = Array.hashCode(Array(1,2,3).toSeq)
                                       ^
Compilation Failed
```

If you're referring to `toArray.hashCode`, that wouldn't work for implementing equality:
```scala
@ Array(1,2,3).hashCode
res0: Int = 1331171857

@ Array(1,2,3).hashCode
res1: Int = 257614494
```

Reply:
Java has a method to compute hash codes of arrays:

```scala
java.util.Arrays.hashCode(array)
```

I'm not sure if `toSeq` is doing that or doing something crazy like converting the array into a vector and doing hash code on that.

Reply:
problem with `Arrays.hashCode` is that it has special implementations for primitives, causing this method to first having to do a check on the type parameter. Nevertheless, the resulting implementation would be much faster. Isn't there a Scala wrapper maybe?